### PR TITLE
Update and add to the GTK and Qt man pages.

### PR DESCRIPTION
**Pull request by Elliria**
This is a major overhaul of the GTK and Qt man pages with these changes:
* Move the date to the top of the document.
* Update the date.
* Move TeX instructions up with the rest of the instructions.
* Update the text in the **NAME** section.
* Reformat the **SYNOPSIS** section.
* Reformat the **DESCRIPTION** section.
* Remove unnecessary information from the **DESCRIPTION** section.
* Update the text in the **DESCRIPTION** section.
* Move the wiki link from the **DESCRIPTION** section to the **SEE ALSO** section.
* Update the introductory text in the **OPTIONS** section.
* Add short command-line switch to the **cutelog** option.
* Rename long command-line switch in the **cutelog** option
* Reformat the **cutelog** option.
* Update the text in the **cutelog** option's description.
* Move the **cutelog** link from the **OPTIONS** section to the **SEE ALSO** section.
* Update the text in the **mouse** option's description.
* Add short command-line switch to the **verbose** option.
* Deprecate the -l command-line switch for the **verbose** option.
* Update the text in the **verbose** option's description.
* Update the text in the **version** option's description.
* Sort the **OPTIONS** section alphabetically.
* Remove man-page-authorship attribution from the **AUTHOR** section.
* Update the text in the **AUTHOR** section.
* Add the **SEE ALSO** section.

Also, only the **Qt** man page contains a reference to **KDialog** and only the **GTK** man page contains a reference to **Zenity**.

### CODE REVIEW

**Comment by Elliria** [comment]
To check out the changes to the man pages in a gentler way than by viewing the raw source code (which is really rough):
1. Visit the [GTK man page](https://github.com/Elliria/autokey/blob/develop/doc/man/autokey-gtk.1) or the [Qt man page](https://github.com/Elliria/autokey/blob/develop/doc/man/autokey-qt.1) in my AutoKey fork.
2. Click the **Download raw file** icon on the right.
3. Open a terminal window in the same directory as the file(s) you downloaded.
4. Type `man ./autokey-gtk.1` or `man ./autokey-qt.1` to see them as actual man pages.

Then, for comparison, you can type `man autokey-gtk` or `man-qt` to see your currently-installed versions of those files.

**Review by josephj11 — COMMENTED** [review]
Suggestions:
29 add "desktop" back in. Automation by itself is too vague. Rest looks fine.

**Review by josephj11 — CHANGES_REQUESTED** [review]
29 add "desktop" back in. Rest looks fine.

**Comment by josephj11** [comment]
FYI: The man page links in your comment above are broken. NBD because I reviewed the changes.

**Comment by Elliria** [comment]
Thanks. I fixed the links. I had deleted my fork and reforked AutoKey and completely forgotten about it when I copied and pasted the links from the other pull request.

**Comment by Elliria** [comment]
There was a missing dash that got added in and **desktop** also got added back in.

**Review by josephj11 — COMMENTED** [review]
This is fine as is, but there are a couple of nits worth picking.

AutoKey is the application and autokey-gtk and -qt are just the front end for it, so it's technically incorrect to say that either of them are the "application".

Maybe "autokey-gtk is a GTK front end for AutoKey, a desktop-automation utility for Linux and X11" would fix this. (with a similar change for the autokey-qt man page)

GNOME is a DTE, but from our perspective, it's a DTE library  one of our front ends uses for its GUI interface. I can and do happily run autokey-gtk under the KDE DTE.

Maybe eliminating " customized for use with GNOME" would fix that.

**Review by josephj11 — COMMENTED** [review]
See my comments on the autokey-gtk man page for how the autokey-qt man page might be slightly improved.

**Comment by Elliria** [comment]
Up until two days ago, I had thought that the updates to the man pages were looking pretty good. Your post, however, pointed out some inconsistencies that have existed for quite some time. As you can imagine, that sent me down a rather deep rabbit hole. Now I realize that those parts of the man pages not only need a complete rewrite, but that we may want to include a new man page as well. At this point, I'm uncertain of the exact wording that the rewrites and/or the new man page should contain and I figured I'd toss the rabbit-hole results into here so that we could put our heads together and figure it out. Without further ado:

You'll find these entries in your your `/usr/bin` directory if you do a search for **autokey** on your system: 
```bash
/usr/bin/autokey
/usr/bin/autokey-gtk
/usr/bin/autokey-qt
/usr/bin/autokey-run
/usr/bin/autokey-shell
```
We're only interested in the **autokey** and **autokey-gtk** and **autokey-qt** executables for the purposes of this discussion. You can manually run any of the three executables by typing either `autokey-gtk` or `autokey-qt` or `autokey` into a terminal window and if you check your process table while each one is running, the name you typed will be shown as its process. This can be misleading, though, because it can trick you into thinking there are three separate and distinct applications that you can run when there are actually only two and the third one is just the hired help.

To see what's actually happening when you run those three commands:
* If you run the `autokey-gtk` command and click on the **Help** entry in the menu, you'll see that you ran the GTK AutoKey application.
* If you run the `autokey-qt` command and click on the **Help** entry in the menu, you'll see that you ran the Qt AutoKey application.
* If you run the `autokey` command and click on the **Help** entry in the menu, you'll see that you either ran the GTK AutoKey application or the Qt AutoKey application (there's no such thing as a generic user-facing AutoKey application that's not customized for either **GTK** or **Qt**).

As you can see from your results of those tests:
* If we choose one of the two applications (**autokey-gtk** or **autokey-qt**), then the application we chose will run.
* If we choose the third application (**autokey**), it chooses one of the two applications for us and runs it on our behalf.

To find out why:
* Take a look in the [doc/man](https://github.com/autokey/autokey/tree/develop/doc/man) directory. You'll see that there's a man page for **autokey-gtk** and another for **autokey-qt**, but there's no man page for **autokey**.
* Take a look inside the [autokey.spec](https://github.com/autokey/autokey/blob/develop/autokey.spec) file. You'll see that the only two AutoKey GUI applications are the [autokey-gtk](https://github.com/autokey/autokey/blob/develop/autokey.spec#L118) application and the [autokey-qt](https://github.com/autokey/autokey/blob/develop/autokey.spec#L126) application. There is no **autokey** application listed.
* Take a look at the [version_upgrading.py](https://github.com/autokey/autokey/blob/develop/lib/autokey/configmanager/version_upgrading.py) file. If you read the section on [lines 164 through 170](https://github.com/autokey/autokey/blob/develop/lib/autokey/configmanager/version_upgrading.py#L164-L170) in that file, you'll see that as of AutoKey 0.95.3, the autostart file was renamed to autokey.desktop and both the GTK and Qt GUIs began sharing a single autokey.desktop autostart entry.

* Type each of these in a terminal window and you'll notice that the first one opens the **GTK** man page for AutoKey, the second one opens thet **Qt** man page for AutoKey, and the third one opens either the **GTK** man or the **Qt** man page depending on the environment you're in when you issue the command:
  * `man autokey-gtk`
  * `man autokey-qt`
  * `man autokey`

What all this means is that we don't have two front ends that influence or affect a single application at all. We have two applications that can be chosen and run separately or can be run by a third application that makes the choice for us.

Reducing all of this to a short summary in the man pages will be a creative and challenging task. Possibly adding a third man page (so that the `man autokey` command will turn up a man page dedicated to explaining the existence of that third application) is also something worth considering.

Thoughts?

**Comment by josephj11** [comment]
TL;DR: It's not a feature, it's a bug (that almost no one cares about).  :)

While I think that the status quo isn't that bad, if I was starting over:

I think this is an initial design flaw.

 I would probably refactor the whole thing so that there is only one user-facing application called autokey. It would then have some way of finding and selecting/running the default front end.

If only one is installed, that's easy, but if both are installed, then there has to be a way to configure the default and probably a CLI flag added to the autokey command to select which front end to run.

This is a bit complicated because the installation process/package needs to know which front end(s) to install because installing both would add dependency bloat that many users might not appreciate (although, compared to containers like snaps and flatpaks, it's pretty small).

As an aside, there is no such thing as the autokey app. (At least in Debian based systems), /usr/bin/autokey is a symlink into the Linux alternatives system which is, in turn, another symlink to the default front end.

If you haven't looked there before, there are a ton of symlinks there for all sorts of things to smooth out how things operate.

So, we actually have one application that is launched indirectly by launching one of its front ends.

This would have to be patched up to work with the refactoring.

**Comment by josephj11** [comment]
Since a refactor isn't in the cards at the moment, if you make any small wording changes ... and whatever went wrong with CHANGELOG.rst, we can merge this and leave elegance for later.

This one might actually qualify as technical debt.

**Comment by Elliria** [comment]
Interesting. The `/usr/bin/autokey` file is a symlink to `/etc/alternatives/autokey`, which is a symlink to `/usr/bin/autokey-qt`. I'll bet that was put there during AutoKey's installation after AutoKey sniffed around and realized this is a KDE environment. I'll also bet that that same symlink would be to `/usr/bin/autokey-gtk` on a GNOME system.

Ah, I see. I'll end up abandoning this pull request and creating a new one as the cleanest and simplest way to avoid that CHANGELOG.rst conflict. And, to prevent them in the future, I should cave in and start using feature branches for changes.

**Comment by Elliria** [comment]
Below are the two man pages in their current condition.

[autokey-gtk.txt](https://github.com/autokey/autokey/files/11604683/autokey-gtk.txt)
[autokey-qt.txt](https://github.com/autokey/autokey/files/11604684/autokey-qt.txt)

I had to rename them with `.txt` extensions to get GitHub to accept them, but they'll work that way. If you grab these two files and then open the directory they're in in a terminal window, you can type either of these to view them:
```
man ./autokey-gtk.txt
```
or:
```
man ./autokey-qt.txt
```

I rewrote the **NAME** and **DESCRIPTION** sections and added a link in each **SEE ALSO** section to the other AutoKey man page.

I'm not happy with the use of "desktop" in the **NAME** and **DESCRIPTION** sections, because AutoKey is much more than that. It can be run invisibly (headless) and make things happen without any visual effect on the desktop.

This is the best I can do so far, but I'm open to suggestions.

**Comment by josephj11** [comment]
It needs something to indicate that it's an end user tool and not a system tool like `at` and `cron`. It's also not like `expect`, ... which is a CLI tool. Saying what it's not is easier that saying what it is. :)

**Comment by Elliria** [comment]
> Saying what it's not is easier that saying what it is. :)

I know. Right? We could go on and on about what it isn't, but coming up with just the right phrasing to convey what it is is an exercise in insanity. What do you think of these two, which are my latest iterations of the man pages? I didn't make it clear that it's not a system tool, but here's hoping:
* [autokey-gtk.txt](https://github.com/autokey/autokey/files/11618765/autokey-gtk.txt)
* [autokey-qt.txt](https://github.com/autokey/autokey/files/11618766/autokey-qt.txt)

As usual, download them and type `man ./autokey-gtk.txt` or `man ./autokey-qt.txt` to view them.

**Comment by josephj11** [comment]
TL;DR: I only know what I don't want. :)

Maybe this should be continued as a discussion aside from this actual PR.

I did a couple of web searches for "desktop automation" and found a lot less than I expected to find. The closest term I found seems to be RPA, robotic process automation, which is also conflated with tools that exercise a GUI app/interface as part of QA infrastructure - automated testing - like what we do with pytest. I'm not sure what we should call it - based on what would most likely be commonly understood. Task automation doesn't do it for me. It's far to broad, imprecise. Bash and all programming languages are task automation tools.

At its heart, AutoKey is a tool which emulates user input, but is also capable of almost anything else because it runs arbitrary user written scripts coded in Python which is a Turing complete language with a very rich set of optional modules available for almost any purpose.

When I originally went looking for something like AutoKey, I was looking for a "macro processor" or language because that is the term I knew from assembly language programming. The only other general purpose tool I found like that (way back then) was m4 which really is a macro processor, but it's so essoteric that I was never able to figure out how to get it to do anything. cpp is like that too, but it's specificly designed for use with the C programming language.

We don't use the term macro in AutoKey because we use script and reserve macro for enhanced tokens that can be added within phrases. But we often encounter users who refer to scripts as macros because that is a more generally understood programming concept.

AutoKey is one application, not two. It happens to have two GUI front ends, one that uses Qt libraries (which are also used by KDE...) and another one that uses GTK libraries (which are also used by GNOME). Again, the choice of one of these front ends is not dependendent on what DTE you are running except in the sense that the respective libraries will only be added to the system as dependencies if they are not already present because your DTE or some other installed application(s) already depend on them. And, except on Arch with its grumpy packaging, you can have both front ends installed and choose which one to run when you start AutoKey.

These front ends can't do much without autokey-common which contains most of the actual functionality of AutoKey (and they won't even install cleanly without autokey-common which they both depend on).

**Comment by Elliria** [comment]
> I did a couple of web searches for "desktop automation" and found a lot less than I expected to find. The closest term I found seems to be RPA, robotic process automation, which is also conflated with tools that exercise a GUI app/interface as part of QA infrastructure - automated testing - like what we do with pytest. I'm not sure what we should call it - based on what would most likely be commonly understood. Task automation doesn't do it for me. It's far to broad, imprecise. Bash and all programming languages are task automation tools.

Perhaps it could be worded this way, to cover all bases:
* desktop- or script-automation utility for GNOME
* desktop- or script-automation utility for KDE
 
> AutoKey is one application, not two. It happens to have two GUI front ends, one that uses Qt libraries (which are also used by KDE...) and another one that uses GTK libraries (which are also used by GNOME).

We've already established that that's not the case at all. [As you saw above](https://github.com/autokey/autokey/pull/867#issuecomment-1566259359), there's no **AutoKey** application, there's the **autokey-gtk** application and the **autokey-qt** application, they each happen to depend upon and share some common code (**autokey-common**), and the **autokey** command is a shortcut for a roll of the dice that chooses a suitable one for us based on our desktop.

Now that you've let the genie out of the bottle by pointing out the symlink that started all that research, I'd rather not be part of circling back to the original descriptions the man pages were using before (or anything like them). I'll bow out of this at this point and close this pull request so you or someone else can take over with a new one.

The other changes that I've made to the updated man pages in the files above (the changes that have nothing to do with the descriptions) really should get added in since they've got nothing to do with the description of AutoKey and would benefit the project by providing explanations of the new command-line options and other useful information that wasn't in the files before.

You (or anyone else) can use the files above and make any changes you like (and remember to update the date at the top of each file) and then use their contents as straight-across replacements for the contents in the [autokey-gtk.1](https://github.com/autokey/autokey/blob/develop/doc/man/autokey-gtk.1) file and the [autokey-qt.1](https://github.com/autokey/autokey/blob/develop/doc/man/autokey-qt.1) file. It should be quick and easy and I'd be happy to help with the steps.

**Comment by josephj11** [comment]
Hey. Didn't mean to offend you (or however that should be categorized). I reserve the right to have strong opinions without that implying that I am right or that I can (or want to) unilaterally decide anything.

This whole arrangement is complicated by the fact that there are usually only two "votes" or opinions - yours and mine with few tie breakers or anything like a majority opinion.

I really do want to get along with you and the rest of the community, but I also don't want to have to walk on egg shells. However, if I have offended you in any way, I am truly sorry. It was not my intention to do so.

You are technically correct about the two separate applications sharing common code. I think I am right in the sense that AutoKey really is one application and because we attempt to have the two front ends be as identical in function as possible (but probably never identical).

On balance, your approach makes more sense until such a time (if ever) that the code is refactored to actually be one application with two front ends. For now, the tail does in fact wag the dog.

Please reinstate your PR and I will merge it. (I could do it, but again, I don't want to override you in any way.)

What about :

- desktop/script-automation utility with a GTK-based graphical interface
- desktop/script-automation utility with a Qt-based graphical interface

or 

- desktop and script-automation utility with a GTK-based graphical interface
- desktop and script-automation utility with a Qt-based graphical interface

?

The fact that there actually is a third program, autokey-run, lends more support to your approach of considering each as separate but related applications.

Autokey-run also has a man page, but it doesn't look like it needs any changes unless you or someone else has improved it and it needs a newer date at the end.

**Comment by Elliria** [comment]
Aside from the questionable egg-shells comment, you haven't offended me. My reason for backing out of it was that I don't want my name on the commits/PR if we were going to go with information that we had already determined to not be true. Also, we had already discussed that this pull request needed to be closed and started over because of the [CHANGELOG.rst](https://github.com/autokey/autokey/blob/develop/CHANGELOG.rst) conflict (my fault for not using feature branches for my commits), so it seemed like the perfect time to do it. Reopening this pull request won't work for that reason, so it has to start over anyway. Last, but not least, the current text above isn't quite right either since it still refers to AutoKey as though it's one application in part of the description.

There are actually two other ways to use AutoKey: **autokey-run** and **autokey-shell**. The **autokey-run** man page will need to be updated after looking through the code to see if any improvements or updates have been made to it, at which point its date will get updated, too. There is no **autokey-shell** man page, which we will need to do something about. I'll create issues for both of those since I just checked and there aren't any. There's a Debian issue about **autokey-shell** and they'll be able to close theirs once we create that.

I've made some more changes to the man pages:
* In the **NAME** section, **desktop and script-automation utility** would be grammatically incorrect because of the missing dash, but **desktop/script-automation utility** will work fine, so I put that into the current version of the man pages.
* In the **DESCRIPTION** section, I took another stab at rewording the text to make it more inclusive of all of the ways in which it can be run.
* The **autokey-run** man page was added to the list of man pages in the **SEE ALSO** section.

Below are two different versions. One set is a detailed version and the other set is a "short and sweet" version:
* [autokey-gtk.txt](https://github.com/autokey/autokey/files/11638841/autokey-gtk.txt)
* [autokey-qt.txt](https://github.com/autokey/autokey/files/11638843/autokey-qt.txt)
* [autokey-gtk-short-and-sweet.txt](https://github.com/autokey/autokey/files/11638840/autokey-gtk-short-and-sweet.txt)
* [autokey-qt-short-and-sweet.txt](https://github.com/autokey/autokey/files/11638842/autokey-qt-short-and-sweet.txt)

As usual, download the files and type these commands to view them as man pages:
* `man ./autokey-qt.txt`
* `man ./autokey-gtk.txt`
* `man ./autokey-gtk-short-and-sweet.txt`
* `man ./autokey-qt-short-and-sweet.txt`

I'm liking the idea of providing details better, but it occurs to me that since we're linking to the documentation, the details could go into the documentation while the man page could just have the basics, and the wording for that can still be tweaked. What do you think?

**Comment by josephj11** [comment]
I like the short and sweet versions much better.

AutoKey is a set of command-line and GUI applications using GTK or Qt that can
       be run in a variety of ways as a desktop/script-automation utility  for  Linux
       and  X11.

Graphics libraries/toolkits and DTEs are independent things. If KDE stopped using Qt (unimaginable in practice), it would still arguably be KDE and Qt would still exist.

KDE is a specific UI/UX and Qt is a tool used for implementing (any) UI/UX.

Same thing for GTK and GNOME.

I thought there was something I was missing - autokey-shell was it.

**Comment by Elliria** [comment]
Yep. I accidentally used GTK instead of GNOME in the most recent files, but changing those to GNOME and KDE can't be done as smoothly as in your example that uses GTK and Qt, so we'll go with that. I've made a slight modification to it to get it to work well with the rest of the text (ending the sentence early and making the original end of it part of the next one. 

I also replaced **options** with all of the short options in the **SYNOPSIS** section (the same way as they are n the **autokey-run** man page if you load that up in a terminal window.

Here are the short-and-sweet ones in their current form:
* [autokey-gtk.txt](https://github.com/autokey/autokey/files/11643389/autokey-gtk.txt)
* [autokey-qt.txt](https://github.com/autokey/autokey/files/11643390/autokey-qt.txt)

**Comment by josephj11** [comment]
The wording is still slightly awkward, but now, what it says is correct. I'd say - good enough for now. I'll merge it when you make a new PR for it.

**Comment by Elliria** [comment]
We may as well keep at it until we get it right since we're so close. Here's another try at it:

Here's the current still-awkward text:

**AutoKey** is a set of command-line and GUI applications using GTK or Qt. It can be run in a variety of ways as a desktop/script-automation utility for Linux and X11 that allows for the automation of virtually any task by responding to typed abbreviations, triggers, and hotkeys. It provides not only a full-featured GUI to make it highly accessible for novices to quickly set up automatic text insertions or replacements, but also a scripting interface to offer the full flexibility and power of the Python language.

Below are some possible alternatives. So far, I like number 8 the best. How about you?

1. **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a desktop/script-automation utility for Linux and X11. **AutoKey** allows for the automation of virtually any task by responding to typed abbreviations, triggers, and hotkeys. It provides a full-featured GUI for setting up automatic text insertions or replacements and a scripting interface for access to the flexibility and power of the Python language.

2. **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a desktop/script-automation utility for Linux and X11. **AutoKey** responds to typed cues by inserting text that you stored as phrases or by inserting the output of scripts that you made with the full flexibility and power of the Python language.

3. **AutoKey** is a set of command-line and GUI applications for Linux and X11 that use GTK or Qt and can be run in a variety of ways as a customizable desktop/script-automation utility. **AutoKey** responds to typed cues by inserting text that you stored as phrases or by inserting the output of scripts that you made with the full flexibility and power of the Python language.

4. **AutoKey** is a set of command-line and GUI applications for Linux and X11 that use GTK or Qt and can be run in a variety of ways as a customizable desktop/script-automation utility. **AutoKey** responds to typed cues by inserting your stored text or the output of your stored scripts that were made with the full flexibility and power of the Python language.

5. **AutoKey** is a set of command-line and GUI applications for Linux and X11 that use GTK or Qt and can be run in a variety of ways as a customizable desktop/script-automation utility. **AutoKey** responds to typed cues by inserting your stored text that can be enhanced with built-in macros or the output of your stored scripts that can be built with the flexibility and power of the Python language and full access to the API.

6. **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a customizable desktop/script-automation utility for Linux and X11. **AutoKey** responds to typed cues by inserting your phrases or the output of your scripts built with the full flexibility and power of the Python language. Additional features include a collection of macros and API calls.

7. **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a customizable desktop/script-automation utility for Linux and X11. **AutoKey** responds to typed cues by inserting your stored text or the output of your stored scripts that were made with the full flexibility and power of the Python language. Macros, API calls, and the full flexibility and power of the Python language are available as additional resources.

8. **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a customizable desktop/script-automation utility for Linux and X11. **AutoKey** stores your custom phrases or scripts and then inserts their contents or output automatically in response to typed cues. Macros, API calls, and the full flexibility and power of the Python language are available as additional resources.

9. Etc.

**Comment by josephj11** [comment]
I'm leaning toward 1 with 5 as a second choice.

In 1, I would leave out 'triggers" because they're part of what an abbreviation is. They're delimiters, not separate entities.

In 5, I would change inserting to emitting. Whether it inserts, overwrites, or does nothing is up to the receiving application/window/field, not to AutoKey.

I've never done it, but I suppose a phrase could consist wholely of  active key tokens like left, right... which might move the text cursor or do other things without emitting anything that would normally be considered as text.

... or by running user-supplied stored Python scripts which can interact with the desktop and system using Autokey's API calls.

----

And while I'm throwing wrenches :) , none of this covers how autokey-run (and probably autokey-shell) does things, but leaving them out here is probably for the best as things are complicted enough already. I don't think I've ever used either of those.

----

Completely unrelated, but I keep forgetting to mention it:

I read that it's more pythonic to use `...` instead of `pass` as placeholder/no op code. Don't remember where I saw it, but I tried it and it works.

**Comment by Elliria** [comment]
> I'm leaning toward 1 with 5 as a second choice.
>
> In 1, I would leave out 'triggers" because they're part of what an abbreviation is. They're delimiters, not separate entities.
>
> In 5, I would change inserting to emitting. Whether it inserts, overwrites, or does nothing is up to the receiving application/window/field, not to AutoKey.

Then I'd say that 1 is the better choice, since **emitting** won't need to be used. How's this?
* **AutoKey** is a set of command-line and GUI applications using GTK or Qt that can be run in a variety of ways as a desktop/script-automation utility for Linux and X11. **AutoKey** allows for the automation of virtually any task by responding to typed abbreviations and hotkeys. It provides a full-featured GUI for setting up automatic text insertions or replacements and a scripting interface for access to the flexibility and power of the Python language.

> I've never done it, but I suppose a phrase could consist wholely of active key tokens like left, right... which might move the text cursor or do other things without emitting anything that would normally be considered as text.

Yes, definitely.

> ... or by running user-supplied stored Python scripts which can interact with the desktop and system using Autokey's API calls.

That would be nice, but would be very limited. The [Run script](https://github.com/autokey/autokey/wiki/Dynamic-Phrases,-Using-Macros-as-placeholders-in-Phrases#run-script) section in the wiki lists several restrictions that could make it either very challenging or impossible.

> And while I'm throwing wrenches :) , none of this covers how autokey-run (and probably autokey-shell) does things, but leaving them out here is probably for the best as things are complicted enough already. I don't think I've ever used either of those.

They both fit within the "command-line" part of the description. Also, one of their man pages has been added to the list of man pages in the **SEE ALSO** section and the other one will be once it's created.

> Completely unrelated, but I keep forgetting to mention it:

> I read that it's more pythonic to use ... instead of pass as placeholder/no op code. Don't remember where I saw it, but I tried it and it works.

That's pretty cool and it does work (I just tried it, too), but I would use it only in very specific circumstances:
* According to [python.org](https://docs.python.org/3/library/constants.html?highlight=ellipsis#Ellipsis), **...** is: "The same as the ellipsis literal “...”. Special value used mostly in conjunction with extended slicing syntax for user-defined container data types. Ellipsis is the sole instance of the types.EllipsisType type."
* According to [python.org](https://docs.python.org/3/reference/simple_stmts.html#the-pass-statement), "pass is a null operation — when it is executed, nothing happens. It is useful as a placeholder when a statement is required syntactically, but no code needs to be executed,"

For the heck of it, I tried doing types on `...` and `pass`:
* If you run the `print(type(...))` command, you'll get the `<class: ellipsis>` result, because the ellipsis is a class.
* If you run the `print(type(pass))` command, you'll get the `SyntaxError: invalid syntax` result, because `pass` is null. It's nothing. 

Last, but not least, since I figure that it's important for anyone who isn't familiar with Python to look up all the parts of any scripts I share, it's easy to look up **python pass**, but **python ...** won't give meaningful results. If a user knows and uses the **ellipsis** name or uses other text to describe them (like **triple dots** or **triple periods** or **three dots** or **three periods** or **dot dot dot** or **dots**), they'll find information, but the first thing I tried was **python ...** when I did the search and others might, too.

**Comment by josephj11** [comment]
As far as ... and pass, I'm a Python newbie, so I was just passing along what I read. Your arguments make sense to me.

Your revision of 1 looks good to me, but if you want to keep discussing other variations, that's fine too.

My bit about scripts was intended to be about normal scripts - not those invoked by the `<script>` macro.

----

I just had a crazy thought: 

It would probably be possible to write an AutoKey script that allowed a user to type or otherwise insert Python code into a file or variable and then run it. I have no idea what this would be good for. It would be challenging to debug anything done this way if it didn't work first try.

A variation of this might be to manually read a module or function from our Modules directory so it was somehow incorporated in the current scope where API calls are available.

I'm not going to pursue this any time soon. I just didn't want the thought to be lost.

**Comment by Elliria** [comment]
> Your arguments make sense to me.

Arguments? I'd like to think of them as observations after exploring the new-to-me Python toy that you had mentioned.
 
> Your revision of 1 looks good to me, but if you want to keep discussing other variations, that's fine too.

I'm good with it. So, should I do another pull request with that one in combination with the other changes that I had made to the man pages?
 
> My bit about scripts was intended to be about normal scripts - not those invoked by the `<script>` macro.

Ah, I hadn't realized that.
 
> I just had a crazy thought:
> 
> It would probably be possible to write an AutoKey script that allowed a user to type or otherwise insert Python code into a file or variable and then run it. I have no idea what this would be good for. It would be challenging to debug anything done this way if it didn't work first try.

That would definitely be doable and might be useful for creating files that are intended to be run by other programs and possibly would also be run by AutoKey upon or after creation. Otherwise, the files wouldn't be necessary since their contents could be fed directly by AutoKey to any processes that would need to be run.
 
> A variation of this might be to manually read a module or function from our Modules directory so it was somehow incorporated in the current scope where API calls are available.

That's an interesting thought. I'm not sure whether or not issues might arise from reading those into AutoKey after they're already been read in by AutoKey or whether there would be any advantage to doing so. It's possible they could be read in with an import statement like: `import foo as bar` and then use `bar` to conjure them. I have no idea whether the same module/library can be read in twice as different objects, but that's worth examining as a possible way to get around some edge-case challenges.

**Comment by josephj11** [comment]
Make a PR (or PRs) that works (any way that makes the most sense to you) and I'll merge it.

I use "arguments" in the [logic/philosophical sense](https://en.wikipedia.org/wiki/Argument), not as in "arguing".

The point of all these weirdnesses is to avoid using import because it doesn't expose the API modules.

What I know about import in general: import tracks what has been imported and if you import something again, it's treated as a no op. IDK what the renaming thing does with that, but I'm pretty sure it works so that you can import the same thing more than once with different "aliases". It probably uses some sort of pointers and a single copy of the imported object.

**Comment by Elliria** [comment]
I went with the revision of 1 and there's now a fresh pull request for both files with the updates we've discussed here.

I also tried importing the same **foo** library into a Python file twice using `import foo as bar` and `import foo as baz` and that worked. I was able to access the contents of **foo** with `bar` and with `baz`. I'm not sure if it would work from within AutoKey as a way to override its use of the library, but it would be worth a try.
